[Macros] In-process plugin server

rintaro opened the pull request:
Separate swift-syntax libs for the compiler and for the library plugins. Compiler communicates with library plugins using serialized messages just like executable plugins.

* `lib/swift/host/compiler/lib_Compiler*.dylib`(source `lib/CompilerSwiftSyntax`): swift-syntax libraries for compiler. Library evolution is disabled.
* Compiler (`ASTGen` and `swiftIDEUtilsBridging`) only depends on `lib/swift/host/compiler` libraries.
* `lib/swift/host/libSwiftInProcPluginServer.dylib` (source `tools/swift-plugin-server/Sources/SwiftInProcPluginServer`): In-process plugin server shared library. This has one `swift_inproc_plugins_handle_message` entry point that receives a message and return the response.

* In the compiler:
  * Add `-in-process-plugin-server-path` front-end option, which specifies the `SwiftInProcPluginServer` shared library path.
  * Remove `LoadedLibraryPlugin`, because all library plugins are managed by `SwiftInProcPluginServer`
  * Introduce abstract `CompilerPlugin` class that has 2 subclasses:
    * `LoadedExecutablePlugin` existing class that represents an executable plugin
    * `InProcessPlugins` wraps `dlopen`ed `SwiftInProcPluginServer` and communicate with it via the `swift_inproc_plugins_handle_message`
  * Unified the code path in `TypeCheckMacros.cpp` and `ASTGen`, the difference between executable plugins and library plugins are now abstracted by `CompilerPlugin`

rintaro commented on lib/ASTGen/Sources/ASTGen/PluginHost.swift:
```diff
@@ -319,7 +319,7 @@ class PluginDiagnosticsEngine {
   ) {
     for diagnostic in diagnostics {
-      self.emit(diagnostic)
+      self.emit(diagnostic, messageSuffix: messageSuffix)
```
We had a bug in executable plugins where ` (from macro '\(macroName)')` suffix wasn't added in the diagnostics from executable plugins 😞

rintaro committed:
[Macros] In-process plugin server

Separate swift-syntax libs for the compiler and for the library plugins.
Compiler communicates with library plugins using serialized messages
just like executable plugins.

* `lib/swift/host/compiler/lib_Compiler*.dylib`(`lib/CompilerSwiftSyntax`):
  swift-syntax libraries for compiler. Library evolution is disabled.
* Compiler (`ASTGen` and `swiftIDEUtilsBridging`) only depends on
  `lib/swift/host/compiler` libraries.
* `SwiftInProcPluginServer`: In-process plugin server shared library.
  This has one `swift_inproc_plugins_handle_message` entry point that
  receives a message and return the response.
* In the compiler
  * Add `-in-process-plugin-server-path` front-end option, which specifies
    the `SwiftInProcPluginServer` shared library path.
  * Remove `LoadedLibraryPlugin`, because all library plugins are managed
    by `SwiftInProcPluginServer`
  * Introduce abstract `CompilerPlugin` class that has 2 subclasses:
    * `LoadedExecutablePlugin` existing class that represents an
      executable plugin
    * `InProcessPlugins` wraps `dlopen`ed `SwiftInProcPluginServer`
  * Unified the code path in `TypeCheckMacros.cpp` and `ASTGen`, the
    difference between executable plugins and library plugins are now
    abstracted by `CompilerPlugin`

hamishknight commented on tools/libSwiftScan/CMakeLists.txt:
```diff
@@ -46,20 +46,20 @@ endif()
 if(SWIFT_BUILD_SWIFT_SYNTAX)
   if(SWIFT_HOST_VARIANT_SDK IN_LIST SWIFT_DARWIN_PLATFORMS)
-    # Ensure that we can find the host shared libraries.
+    # Ensure that we can find the shaered swiftsyntax libraries.
```
```suggestion
    # Ensure that we can find the shared swiftsyntax libraries.
```

rintaro commented on lib/AST/PluginRegistry.cpp:
```diff
@@ -44,46 +44,97 @@ PluginRegistry::PluginRegistry() {
+}
+
+llvm::Expected<InProcessPlugins *>
+InProcessPlugins::create(const char *serverPath) {
```
No. Changed to return `unique_ptr` 👍

rintaro commented on lib/AST/PluginRegistry.cpp:
```diff
@@ -44,46 +44,97 @@ PluginRegistry::PluginRegistry() {
+  assert(!receivedResponse.empty() &&
+         "waitForNextMessage() called without response data.");
+  auto message = receivedResponse;
+  receivedResponse = "";
```
👍 
```
SWIFT_DEFER { receivedResponse = ""; };
return std::move(receivedResponse);
```

rintaro commented on cmake/modules/AddPureSwift.cmake:
```diff
@@ -280,12 +254,27 @@ function(add_pure_swift_host_library name)
-        Swift_MODULE_NAME ${name}
-        Swift_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
-    set(module_file "${CMAKE_CURRENT_BINARY_DIR}/${name}.swiftmodule")
+    set(module_dir "${CMAKE_CURRENT_BINARY_DIR}/modules")
```
Fragile modules are now created in `${CMAKE_CURRENT_BINARY_DIR}/modules`. This is to avoid PR testing failures in incremental bots. This PR changes `_CompilerRegexParser` to fragile, if we create `${CMAKE_CURRENT_BINARY_DIR}/_CompilerRegexParser.swiftmodule` , subsequent PR testings will be confused because `-I ${CMAKE_CURRENT_BINARY_DIR}` is added by CMake. Pre-patch https://github.com/apple/swift/pull/73799 can avoid that, but just in case.

rintaro commented on test/Macros/macro_expand.swift:
```diff
@@ -629,8 +629,8 @@ struct HasNestedType {
 @freestanding(expression)
 macro missingMacro() = #externalMacro(module: "MacroDefinition", type: "BluhBlah")
-// expected-warning@-1 {{external macro implementation type 'MacroDefinition.BluhBlah' could not be found for macro 'missingMacro()'; 'MacroDefinition.BluhBlah' could not be found in library plugin '}}
+// FIXME: xpected-warning@-1 {{external macro implementation type 'MacroDefinition.BluhBlah' could not be found for macro 'missingMacro()'; 'MacroDefinition.BluhBlah' could not be found in library plugin '}}
```
This is because the plugin message facilities (i.e. executable plugins) doesn't support "resolve macro type" at this point. Maybe we can introduce `case resolveMacroType(module: String, type: String)` plugin message.

rintaro commented on tools/libSwiftScan/CMakeLists.txt:
```diff
@@ -46,20 +46,20 @@ endif()
     set_property(
       TARGET libSwiftScan
-      APPEND PROPERTY INSTALL_RPATH "@loader_path/../host")
+      APPEND PROPERTY INSTALL_RPATH "@loader_path/compiler")
```
https://github.com/apple/swift/commit/1e93fb34f821941ae533a46e17af9b8d0eb44684
Was there a reason for `@loader_path/../host` instead of `@loader_path`?

rintaro replied:
Yes, it's symlinked in `lib/swift/linux` etc.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please test Windows

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test Windows

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test Windows

kabiroberai commented:
:wave: does this PR remove support for the `loadPluginLibrary` approach entirely? Worth noting that we're relying on this mechanism in https://github.com/apple/swift/pull/73031 in order to enable support for WebAssembly-based plugins. We could move these in-process but I think it would be preferable to keep third-party Wasm plugins out-of-process for defence-in-depth/isolation reasons (especially if we eventually want to execute them with JIT), so IMO it's worth keeping the library plugin mechanism for that. Also, IIUC one of the ways we could improve macro performance is to reuse a single plugin library/executable across `swiftc` invocations. Would this PR preclude that change for library-based plugins?

rintaro commented:
> does this PR remove support for the loadPluginLibrary approach entirely?

I'm not 100% sure what `loadPluginLibrary` approach you mean, but either it's about  `-load-plugin-library` compiler option or `loadPluginLibrary` plugin message, this PR doesn't remove them. 

Since we need to keep supporting "in-process" plugins (i.e. `-load-plugin-library` and `-plugin-path` compiler options) for various reasons, we've been suffered by some limitations from that. The main intent of this PR is that separating swift-syntax libraries for the compiler and swift-syntax libraries for those "in-process" plugins. By that
* We can unify the code path between in-process and out-of-process plugins
* We can use different swift-syntax between the compiler and the in-process plugins. That let us compiling swift-syntax for the compiler _without_ library-evolution, for the compiler performance

This PR doesn't have any intent of removing or deprecating out-of-process plugin mechanism.

> IIUC one of the ways we could improve macro performance is to reuse a single plugin library/executable across swiftc invocations. Would this PR preclude that change for library-based plugins?

Yes, reusing plugins across multiple `swift-frontend` invocations is probably what we need to tackle on next. This PR does _not_ preclude that approach.

I'm not sure I entirely understand your concerns. Let me know if you still have any questions!

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

kabiroberai commented:
Ah I was under the impression that this removed support for external plugin servers, but looks like I was confused. Thanks for the clarification!

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test Linux

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test Linux

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
@swift-ci Please Test Source Compatibility

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

etcwilde commented on stdlib/public/Windows/CMakeLists.txt:
```diff
@@ -4,5 +4,8 @@ add_swift_target_library(swiftWinSDK ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} IS_SDK_
-    INSTALL_IN_COMPONENT sdk-overlay)
-
+    INSTALL_IN_COMPONENT sdk-overlay
+    # Workaround for the issue swiftWinSDK.dll in the built directory might be being used
```
This sounds like a build-system architectural issue. Nothing that we just-built should be running. Will keep my eyes open for this as I continue pushing through the build. 🤔

rintaro replied:
Yes. This is a horrible workaround. But I haven't been able to find a way to avoid it.

compnerd commented:
@swift-ci please build toolchain Windows platform

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please build toolchain Windows

ahoppen commented on lib/ASTGen/Sources/ASTGen/PluginHost.swift:
```diff
@@ -378,6 +374,26 @@ class PluginDiagnosticsEngine {
 }
 
+extension String {
+  /// Retrieve the base name of a string that represents a path, removing the
+  /// directory.
+  var basename: String {
+    guard
+      let lastSlash = lastIndex(where: {
+        #if os(iOS) || os(macOS) || os(tvOS) || os(watchOS) || os(visionOS) || os(Android) || os(Linux)
+        ["/"].contains($0)
+        #else
+        ["/", "\\"].contains($0)
+        #endif
+      })
+    else {
+      return self
+    }
+
+    return String(self[index(after: lastSlash)...])
+  }
```
I don’t think this is used anywhere.

rintaro replied:
This is moved from deleted [lib/ASTGen/Sources/ASTGen/SourceManager+MacroExpansionContext.swift](https://github.com/apple/swift/pull/73725#diff-837202e3d3c3d5f4d1db10aff9aebdabdae57a6d16e58fc184c06266eb542844)
And is used at https://github.com/apple/swift/blob/03b0f0cc57feb18246348168fb2f937e95496277/lib/ASTGen/Sources/ASTGen/PluginHost.swift#L394-L397

ahoppen replied:
Ah, that’s why I didn’t see it.

rintaro commented on lib/AST/PluginRegistry.cpp:
```diff
@@ -44,46 +44,96 @@ PluginRegistry::PluginRegistry() {
+    llvm::dbgs() << "<-(plugin:0) " << receivedResponse << "\n";
+  }
+
+  assert(!receivedResponse.empty() && "received empty response()");
```
Yes. Every plugin message is a JSON.

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

ahoppen commented on include/swift/AST/PluginRegistry.h:
```diff
@@ -25,26 +26,114 @@
+  std::function<void(void)> cleanup;
+
+  /// Callbacks to be called when the connection is restored.
+  llvm::SmallVector<std::function<void(void)> *, 0> onReconnect;
```
Fair enough 👍🏽

rintaro committed:
[Macros] Infer in-process plugin server in swift-frontend

Not all driver can send '-in-process-plugin-server-path'. To keep
existing '-plugin-path' working, infer default server path in the
frontend.

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please Build Toolchain macOS

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-syntax/pull/2659
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please Build Toolchain macOS

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/8855
https://github.com/apple/swift-driver/pull/1616
https://github.com/apple/swift-installer-scripts/pull/304
@swift-ci Please Build Toolchain Windows
